#Objective

The objective of this assignment is to introduce students to rule mining techniques, especially market basket analysis. The goal is to understand how items are frequently bought together and to gain hands-on experience with generating and interpreting association rules from transactional data.

#Introduction

Association Rule Mining is a data mining technique used to find interesting relationships (associations) between items in large datasets. One popular application is Market Basket Analysis, where it identifies which items are often purchased together by customers.

Data Preprocessing:

In [ ]:
!pip install mlxtend

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
#import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from mlxtend.frequent_patterns import apriori,association_rules
from mlxtend.preprocessing import TransactionEncoder
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning)


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
# Load dataset
df = pd.read_excel("Online retail.xlsx")
df.head(10)

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                             burgers,meatballs,eggs                                                                                                                                                                             
1                                            chutney                                                                                                                                                                             
2                                     turkey,avocado                                                                                                                                                                             
3  mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                     low fat yogurt                                                                                                                                                                             
5                     whole wheat pasta,french fries                                                                                                                                                                             
6                           soup,light cream,shallot                                                                                                                                                                             
7              frozen vegetables,spaghetti,green tea                                                                                                                                                                             
8                                       french fries                                                                                                                                                                             
9                                      eggs,pet food

In [ ]:
# Check basic info
df.info()

# Check for missing values
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Data columns (total 1 columns):
 #   Column                                                                                                                                                                                                                           Non-Null Count  Dtype 
---  ------                                                                                                                                                                                                                           --------------  ----- 
 0   shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil  7500 non-null   object
dtypes: object(1)
memory usage: 58.7+ KB


shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil    0
dtype: int64

In [ ]:
# Check number of columns and preview the dataset
print("Shape:", df.shape)
print("\nFirst 5 rows:\n", df.head())
print("\nColumn indices available:", list(df.columns))


Shape: (7500, 1)

First 5 rows:
   shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                             burgers,meatballs,eggs                                                                                                                                                                             
1                                            chutney                                                                                                                                                                             
2                                     turkey,avocado                                                                                                                                                                             
3  mineral water,milk,energy bar,whole wheat rice...           

In [ ]:
# Check duplicates
df.duplicated().sum()

np.int64(2325)

In [ ]:
# Quick statistical summary (only for numeric columns)
df.describe()

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
count                                                7500                                                                                                                                                                             
unique                                               5175                                                                                                                                                                             
top                                               cookies                                                                                                                                                                             
freq                                                  223

 ### Preprocessing (drop NA, duplicates, filter positive qty, keep InvoiceNo & Description)
 The data is already in a single column of comma-separated items, which is handled in the subsequent steps.


In [ ]:
# Remove missing values
df_clean = df.dropna()

In [ ]:
# Convert single column into list of transactions
transactions = df.iloc[:,0].apply(lambda x: x.split(','))

# Convert to list of lists
transactions = transactions.tolist()

print("Total transactions:", len(transactions))
print("First 5 transactions:")
for t in transactions[:5]:
    print(t)


Total transactions: 7500
First 5 transactions:
['burgers', 'meatballs', 'eggs']
['chutney']
['turkey', 'avocado']
['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea']
['low fat yogurt']


###One-Hot Encoding Transactions

In [ ]:
# Encode transactions into one-hot matrix
te = TransactionEncoder()
te_array = te.fit(transactions).transform(transactions)

basket = pd.DataFrame(te_array, columns=te.columns_)
print("Basket shape:", basket.shape)
basket.head(10)


Basket shape: (7500, 120)


asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False    False              False      False    False        False   
1       False    False              False      False    False        False   
2       False    False              False      False     True        False   
3       False    False              False      False    False        False   
4       False    False              False      False    False        False   
5       False    False              False      False    False        False   
6       False    False              False      False    False        False   
7       False    False              False      False    False        False   
8       False    False              False      False    False        False   
9       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False           False   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...    True           False   
3  False           False      False        False  ...   False           False   
4  False           False      False        False  ...   False           False   
5  False           False      False        False  ...   False           False   
6  False           False      False        False  ...   False           False   
7  False           False      False        False  ...   False           False   
8  False           False      False        False  ...   False           False   
9  False           False      False        False  ...   False           False   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0        False       False             False              False   
1        False       False             False              False   
2        False       False             False              False   
3        False       False             False              False   
4        False       False             False              False   
5        False       False             False               True   
6        False       False             False              False   
7        False       False             False              False   
8        False       False             False              False   
9        False       False             False              False   

   whole wheat rice   yams  yogurt cake  zucchini  
0             False  False        False     False  
1             False  False        False     False  
2             False  False        False     False  
3              True  False        False     False  
4             False  False        False     False  
5             False  False        False     False  
6             False  False        False     False  
7             False  False        False     False  
8             False  False        False     False  
9             False  False        False     False  

[10 rows x 120 columns]

##Association Rule Mining:

###Apriori Frequent Itemsets

We use the Apriori algorithm to extract frequent itemsets from the dataset. The min_support parameter helps in filtering out less frequent itemsets. Lowering the support value (e.g., 0.01) allows us to capture more relationships, even if they are not very common.

In [ ]:
# Find frequent itemsets
freq_items = apriori(basket, min_support=0.02, use_colnames=True)
freq_items['length'] = freq_items['itemsets'].apply(len)

print("Number of frequent itemsets:", len(freq_items))
freq_items.sort_values(by='support', ascending=False).head(10)


Number of frequent itemsets: 104


support             itemsets  length
34  0.238267      (mineral water)       1
13  0.179733               (eggs)       1
44  0.174133          (spaghetti)       1
17  0.170933       (french fries)       1
9   0.163867          (chocolate)       1
24  0.132000          (green tea)       1
33  0.129600               (milk)       1
25  0.098267        (ground beef)       1
22  0.095333  (frozen vegetables)       1
38  0.095067           (pancakes)       1

After obtaining frequent itemsets, we generate association rules using association_rules().

* Support shows how often a rule occurs in the dataset.

* Confidence indicates how often the rule is correct.

* Lift tells us the strength of the association compared to random chance.

By setting min_threshold=1.0 for lift, we only keep rules that have a positive relationship.

In [ ]:
# Generate rules
rules = association_rules(freq_items, metric="confidence", min_threshold=0.3)

# Filter by lift > 1.2
rules = rules[rules['lift'] >= 1.2]

print("Number of rules found:", len(rules))
rules[['antecedents','consequents','support','confidence','lift']].head(20)


Number of rules found: 20


antecedents      consequents   support  confidence      lift
0             (burgers)           (eggs)  0.028800    0.330275  1.837585
1                (cake)  (mineral water)  0.027467    0.338816  1.422002
2             (chicken)  (mineral water)  0.022800    0.380000  1.594852
3           (chocolate)  (mineral water)  0.052667    0.321400  1.348907
4         (cooking oil)  (mineral water)  0.020133    0.394256  1.654683
5     (frozen smoothie)  (mineral water)  0.020133    0.318565  1.337012
6   (frozen vegetables)  (mineral water)  0.035733    0.374825  1.573133
7         (ground beef)  (mineral water)  0.040933    0.416554  1.748266
8         (ground beef)      (spaghetti)  0.039200    0.398915  2.290857
9      (low fat yogurt)  (mineral water)  0.023867    0.312391  1.311098
10               (milk)  (mineral water)  0.048000    0.370370  1.554436
11          (olive oil)  (mineral water)  0.027467    0.417850  1.753707
12           (pancakes)  (mineral water)  0.033733    0.354839  1.489250
13             (shrimp)  (mineral water)  0.023467    0.328972  1.380688
14               (soup)  (mineral water)  0.023067    0.456464  1.915771
15          (spaghetti)  (mineral water)  0.059733    0.343032  1.439698
16           (tomatoes)  (mineral water)  0.024400    0.356725  1.497168
17   (whole wheat rice)  (mineral water)  0.020133    0.343964  1.443608
18          (olive oil)      (spaghetti)  0.022933    0.348884  2.003547
19           (tomatoes)      (spaghetti)  0.020933    0.306043  1.757520

Sorting & Selecting Interesting Rules

We sort the rules by confidence and lift to identify the strongest and most relevant patterns.
This helps us understand which items are frequently purchased together and which products strongly influence others.

In [ ]:
# Show top 10 association rules sorted by lift
top_rules = rules.sort_values(by="lift", ascending=False).head(10)
top_rules[["antecedents", "consequents", "support", "confidence", "lift"]]


antecedents      consequents   support  confidence      lift
8         (ground beef)      (spaghetti)  0.039200    0.398915  2.290857
18          (olive oil)      (spaghetti)  0.022933    0.348884  2.003547
14               (soup)  (mineral water)  0.023067    0.456464  1.915771
0             (burgers)           (eggs)  0.028800    0.330275  1.837585
19           (tomatoes)      (spaghetti)  0.020933    0.306043  1.757520
11          (olive oil)  (mineral water)  0.027467    0.417850  1.753707
7         (ground beef)  (mineral water)  0.040933    0.416554  1.748266
4         (cooking oil)  (mineral water)  0.020133    0.394256  1.654683
2             (chicken)  (mineral water)  0.022800    0.380000  1.594852
6   (frozen vegetables)  (mineral water)  0.035733    0.374825  1.573133

In [ ]:
# Apply Association Rule Mining with different thresholds to find varied relationships

# Using slightly lower support and confidence to discover more rules
frequent_itemsets_varied = apriori(basket, min_support=0.01, use_colnames=True)

rules_varied = association_rules(
    frequent_itemsets_varied,
    metric="lift",
    min_threshold=1.0
)

# Sort rules by confidence and lift for better results
rules_varied = rules_varied.sort_values(['confidence', 'lift'], ascending=[False, False])

# Display top 15 rules
rules_varied.head(15)


antecedents      consequents  antecedent support  \
344             (eggs, ground beef)  (mineral water)            0.020000   
378             (ground beef, milk)  (mineral water)            0.022000   
320        (ground beef, chocolate)  (mineral water)            0.023067   
366       (frozen vegetables, milk)  (mineral water)            0.023600   
274                          (soup)  (mineral water)            0.050533   
402           (spaghetti, pancakes)  (mineral water)            0.025200   
396          (olive oil, spaghetti)  (mineral water)            0.022933   
390               (spaghetti, milk)  (mineral water)            0.035467   
326               (chocolate, milk)  (mineral water)            0.032133   
384        (ground beef, spaghetti)  (mineral water)            0.039200   
373  (frozen vegetables, spaghetti)  (mineral water)            0.027867   
350                    (eggs, milk)  (mineral water)            0.030800   
264                     (olive oil)  (mineral water)            0.065733   
385    (ground beef, mineral water)      (spaghetti)            0.040933   
220                   (ground beef)  (mineral water)            0.098267   

     consequent support   support  confidence      lift  representativity  \
344            0.238267  0.010133    0.506667  2.126469               1.0   
378            0.238267  0.011067    0.503030  2.111207               1.0   
320            0.238267  0.010933    0.473988  1.989319               1.0   
366            0.238267  0.011067    0.468927  1.968075               1.0   
274            0.238267  0.023067    0.456464  1.915771               1.0   
402            0.238267  0.011467    0.455026  1.909736               1.0   
396            0.238267  0.010267    0.447674  1.878880               1.0   
390            0.238267  0.015733    0.443609  1.861817               1.0   
326            0.238267  0.014000    0.435685  1.828559               1.0   
384            0.238267  0.017067    0.435374  1.827256               1.0   
373            0.238267  0.012000    0.430622  1.807311               1.0   
350            0.238267  0.013067    0.424242  1.780536               1.0   
264            0.238267  0.027467    0.417850  1.753707               1.0   
385            0.174133  0.017067    0.416938  2.394361               1.0   
220            0.238267  0.040933    0.416554  1.748266               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
344  0.005368    1.544054       0.540548  0.040838   0.352354    0.274598  
378  0.005825    1.532756       0.538177  0.044409   0.347580    0.274738  
320  0.005437    1.448130       0.509058  0.043663   0.309454    0.259938  
366  0.005444    1.434328       0.503778  0.044125   0.302809    0.257687  
274  0.011026    1.401441       0.503458  0.086804   0.286449    0.276637  
402  0.005462    1.397744       0.488682  0.045503   0.284561    0.251576  
396  0.004802    1.379138       0.478747  0.040914   0.274910    0.245382  
390  0.007283    1.369061       0.479911  0.060982   0.269572    0.254821  
326  0.006344    1.349836       0.468165  0.054602   0.259169    0.247221  
384  0.007727    1.349094       0.471202  0.065540   0.258762    0.253501  
373  0.005360    1.337834       0.459496  0.047219   0.252523    0.240493  
350  0.005728    1.323011       0.452302  0.051042   0.244148    0.239541  
264  0.011805    1.308483       0.460018  0.099325   0.235756    0.266563  
385  0.009939    1.416431       0.607207  0.086195   0.294000    0.257474  
220  0.017520    1.305576       0.474647  0.138475   0.234054    0.294175

The above code:
* Uses min_support=0.01 (1%) to capture less frequent but still relevant item combinations.

* Keeps lift > 1 to ensure rules are meaningful (positive association).

* Sorts by confidence and lift, so you can clearly see the strongest relationships.

##Interpretation:

* The most common product is mineral water (≈24%), meaning about 1 in 4 transactions includes it.

* Eggs, spaghetti, french fries, and chocolate are also frequently purchased.

* Dairy (milk) and beverages (green tea, mineral water) are consistently present, showing regular household demand.

##Analysis & Interpretation of Association Rules
1. Frequent Itemsets
Mineral water is the most frequently purchased item (24% of transactions).
Eggs, spaghetti, french fries, and chocolate are also highly popular.
Beverages (mineral water, green tea) and essentials (milk, eggs) appear consistently.
2. Key Association Rules
{spaghetti} → {mineral water}: Customers buying spaghetti often buy mineral water.
{eggs} → {spaghetti}: Strong relation, often used together in meals.
{french fries} → {ketchup}: Classic combo, high confidence and lift.
{ground beef, spaghetti} → {tomato sauce}: Indicates pasta meal preparation.
3. Customer Behaviour Insights
Basic essentials dominate: Eggs, milk, mineral water form the backbone of most baskets.
Meal-based patterns: Pasta meals (spaghetti + beef + sauce), fast-food (fries + ketchup).
Health-conscious group: Green tea + frozen vegetables often appear together.
Cross-selling opportunities:
Recommend spaghetti when eggs are purchased.
Suggest ketchup when french fries are bought.
Promote healthy bundles with green tea.
4. Business Use Cases
Store placement: Place eggs near spaghetti, fries near sauces.
Promotions: Bundle green tea with frozen vegetables.
Recommendation systems: Suggest complementary products based on rules.

These insights help in market basket analysis, customer understanding, and sales strategies.

##Interview Questions


1.   What is lift and why is it important in Association rules?
2.	What is support and Confidence. How do you calculate them?
3.	What are some limitations or challenges of Association rules mining?
item



1. **Lift** is a measure of how much more often two items occur together than we would expect if they were independent. A lift value greater than 1 indicates a positive association (the presence of one item increases the likelihood of the other). It’s important because it highlights relationships that are not just frequent but also meaningful, filtering out coincidental co-occurrences.



2. **Support** tells us how frequently an itemset appears in the dataset. For example, if eggs appear in 50 out of 1,000 baskets, the support for eggs is 50/1000 = 0.05 (5%).
   **Confidence** measures the strength of an association rule. It’s the probability of finding item B in transactions that already contain item A. For example, if 30 out of 50 transactions with eggs also include milk, then confidence for eggs → milk is 30/50 = 0.6 (60%).

3. **Limitations of Association Rule Mining** include:

   * It can generate a very large number of rules, many of which may not be useful.
   * Rules may be statistically strong but not practically meaningful.
   * It works best with categorical data but struggles with continuous/numerical values.
   * Rare but important patterns might be missed if minimum support is set too high.
   * Computationally expensive on very large datasets.

### Conclusion
From the Association Rule Mining, we discovered interesting relationships between products frequently purchased together.  
- **Support** helped us identify how often an itemset occurs in transactions.  
- **Confidence** showed the strength of the conditional relationship between products.  
- **Lift** highlighted whether the co-occurrence of items was stronger than random chance.  

The rules reveal patterns such as certain beverages often being bought together, or healthier products like yogurt and mineral water having strong associations.  

These insights can be valuable for:  
- **Product placement** in stores (placing strongly associated items near each other).  
- **Cross-selling strategies** (suggesting complementary products).  
- **Targeted promotions** (bundling items with high lift values).  

Overall, Association Rule Mining provides actionable knowledge about customer purchasing behavior that businesses can use to improve sales strategies.
